ambry-router/src/main/java/com.github.ambry.router/SimpleOperationTracker.java

```diff
@@ -357,20 +388,40 @@ public ReplicaId next() {
 
   private boolean hasFailed() {
     boolean hasFailed;
-    if (routerOperation == RouterOperation.PutOperation && routerConfig.routerPutUseDynamicSuccessTarget) {
+    if (routerOperation == RouterOperation.PutOperation && routerConfig.routerPutUseDynamicSuccessTarget
+        && inFlightReplicaType == ReplicaType.DISK_BACKED) {
       hasFailed = totalReplicaCount - failedCount < Math.max(totalReplicaCount - 1,
           routerConfig.routerPutSuccessTarget + disabledCount);
     } else {
-      hasFailed = (totalReplicaCount - failedCount) < successTarget || hasFailedOnNotFound();
+      // if one type of replica has a lower success target, we can only call it a complete failure when there are fewer
+      // than that target left to check. This could be optimized by saving the replica type of the remaining replicas
+      // to try and choosing the success target based on their type.
+      int minSuccessTarget = cloudReplicasPresent ? Math.min(diskSuccessTarget, cloudSuccessTarget) : diskSuccessTarget;
+      hasFailed = (totalReplicaCount - failedCount) < minSuccessTarget || hasFailedOnNotFound();
```

Comment:
I'm having trouble following this logic.  Assuming there is a cloud replica present, totalReplicaCount = 13, minSuccessTarget will be 1, in which case failedCount has to equal totalReplicaCount to declare failure.  But if the cloud replica has failed, we will still require diskSuccessTarget successes among disk replicas - assume that is 2.  Then the most failures we can tolerate is 11 (10 disk + 1 cloud).  With failedCount = 12, hasFailed = (13-12 < 1) will be false but should be true, no?

Reply:
I guess this is a little rough to use the min. The reason I did that was to account for the case where the 13th replica was a cloud replica that could still satisfy the request with a success target of 1. However, this has the downside of wasting a request if that last replica was a disk replica.

To address this, I could do what the comment suggests and figure out the replica types of the remaining replicas. To avoid iterating through the entire list every time hasFailed is called, I could record the index of the last cloud replica and the last disk replica in the pool. If we are past the last replica of a certain type, switch to the other success target. Or, in this method, compare against the max success target and then if that returns true, make a second comparison by inspecting the remaining replicas in the pool (this should be a relatively smaller # of items to iterate through assuming success targets are small).

Do either of these approaches sound better?

Reply:
Sounds like 1st approach is more efficient.

Reply:
In addition, let's also account for PUT operation when this is an Azure frontend. I feel like we need an extra `else if` for this case because the first `if` only applies to DiskReplica in regular dc(s).

Reply:
Hi @jsjtzyy, for the put operation in azure, what condition were you thinking of where the logic will be different from the fallback policy? I feel like there is no concept of disabled for cloud replicas.

Reply:
For this condition I added a check in `hasFailedOnNotFound` to always return false on put operations. The condition would have returned false anyways given how we instantiate the operationtracker in PutOperation, but this makes it clearer.

For the failure threshold I changed the logic around a bit by counting how many replicas of each type are left. Using the success count for each type and this number, we can tell if a success is still possible or not for each type without requiring the hacky min logic